src/lib/y2partitioner/actions/move_partition.rb

```diff
@@ -0,0 +1,344 @@
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "yast/i18n"
+require "yast2/popup"
+require "y2partitioner/dialogs/partition_move"
+require "y2partitioner/device_graphs"
+require "y2partitioner/ui_state"
+
+module Y2Partitioner
+  module Actions
+    # Action for moving a partition
+    #
+    # To move a partition means to place it at the beginning or at the end of the
+    # previous or next adjacent free space, respectively.
+    class MovePartition
+      include Yast::I18n
+
+      # Constructor
+      #
+      # @param partition [Y2Storage::BlkDevice]
+      def initialize(partition)
+        textdomain "storage"
+
+        @partition = partition
+        UIState.instance.select_row(partition.sid)
+      end
+
+      # Checks whether it is possible to move the partition, and if so, the action is performed.
+      #
+      # @note An error popup is shown when the partition cannot be moved. In case the partition
+      #   can be moved, a dialog is presented to select whether to move forward or backward.
+      #
+      # @return [Symbol] :finish if the action is performed; :back or dialog result otherwise.
+      def run
+        return :back unless validate
+
+        answer = move_partition_question
+        return answer unless answer == :ok
+
+        move_partition
+        :finish
+      end
+
+    private
+
+      # @return [Y2Storage::BlkDevice]
+      attr_reader :partition
+
+      # @return [Symbol] :forward, :backward
+      attr_reader :movement
+
+      # Checks whether the partition can be moved
+      #
+      # @note When the partition cannot be moved, it shows a popup with the reason,
+      #   see {#error}.
+      #
+      # @return [Boolean] true if the partition can be moved; false otherwise.
+      def validate
+        return true unless error
+
+        Yast2::Popup.show(error, headline: :error)
+        false
+      end
+
+      # Why the partition cannot be moved
+      #
+      # @note A partition can be moved if is not an extended partition and it does
+      #   not exist on disk and there is adjancent free space where to move it. In
+      #   some cases, the move action could be called over a device that is not a
+      #   partition (i.e., when the selected element on a table is not a partition).
+      #   For that reason, the type of the device must be checked.
+      #
+      # @return [String, nil] nil when there is no error and the partition can be moved.
+      def error
+        device_type_error ||
+          extended_partition_error ||
+          existing_partition_error ||
+          no_space_error
+      end
+
+      # Error when a no partition device is tried to be moved
+      #
+      # @return [String, nil] nil the device is a partition
+      def device_type_error
+        return nil if partition.is?(:partition)
+
+        _("Hard disks, BIOS RAIDs and multipath\n" \
```

Comment:
this looks like a bit insuficient to the above check. Can we write only that "Only partition can be moved."?

Reply:
Well, I basically copied the text from the old partitioner.